mitodl/mitxpro
# Front-end coupon creation

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes #107 

#### What's this PR do?
Front-end coupon creation page and form.

#### How should this be manually tested?
- If you haven't already, create products for your courses, runs, and programs:
```
for courserun in CourseRun.objects.all():
    Product.objects.create(object_id=courserun.id, content_type=ContentType.objects.get(model="courserun"))

for course in Course.objects.all():
    Product.objects.create(object_id=course.id, content_type=ContentType.objects.get(model="course"))

for program in Program.objects.all():
    Product.objects.create(object_id=program.id, content_type=ContentType.objects.get(model="program"))
```
- Start a company or two:
```
Company.objects.create(name="Acme Corporation")
Company.objects.create(name="SpaceX")
```
- Go to /`ecommerce/admin/coupons`
- Try creating both single-use and promo coupons
- The following fields are mandatory for single-use coupons but not promo coupons:
  - Number of coupons
  - Payment type
  - Payment transaction
- The following field is mandatory for promo coupons but not single-use coupons:
  - Coupon code
- The following fields are mandatory for all coupons:
  - Name
  - Activation date
  - Expiration date
  - Products (at least 1 selected)
- Anything other than letters, digits and underscores should not be accepted for coupon name or coupon code.
- After successfully submitting a single-use coupon, the form should disappear;  you should see a link to download your coupon codes and a button to display the form again.  Clicking the link should download a csv file called `coupon_codes_<coupon_name>.csv` containing the number of codes specified on the form.
- After successfully submitting a promo coupon, the form should disappear;  you should see a message that your coupon was created, and a button to display the form again. 


#### Screenshots 
#### Basic form:
<img width="753" alt="Screen Shot 2019-04-11 at 6 56 17 PM" src="https://user-images.githubusercontent.com/187676/55998507-851da480-5c8c-11e9-96ba-4117eea24a2e.png">

--------------

#### Sample validation error on submission:
<img width="226" alt="Screen Shot 2019-04-11 at 7 02 53 PM" src="https://user-images.githubusercontent.com/187676/55998496-77681f00-5c8c-11e9-9041-1fa4c7963f9c.png">

--------------

#### Date picker
<img width="466" alt="Screen Shot 2019-04-10 at 12 00 13 PM" src="https://user-images.githubusercontent.com/187676/55894536-4272a400-5b88-11e9-92be-975cb7c432be.png">


#### Success message for promo coupon, form is hidden
<img width="452" alt="Screen Shot 2019-04-11 at 6 57 24 PM" src="https://user-images.githubusercontent.com/187676/55998355-e7c27080-5c8b-11e9-9eb8-941068ba36d6.png">

-------------

#### Success message / download link for single-use coupons, form is hidden:
<img width="383" alt="Screen Shot 2019-04-11 at 6 56 35 PM" src="https://user-images.githubusercontent.com/187676/55998419-1f311d00-5c8c-11e9-816b-86ae88d5d373.png">
